### PR TITLE
Use resource manager to download meta files. Fixes #4234

### LINKED ISSUE #4234: [Image Classification API] TensorFlow exception triggered: input ended unexpectedly in the middle of a field
### System information

- **OS version/distro**: Windows 10
- **.NET Version (eg., dotnet --info)**: .NET Core 2.2

### Issue

- **What did you do?**

Tried to train an image classification DNN model using the Image Classification API on the [Intel Image Classification](https://www.kaggle.com/puneet6060/intel-image-classification) dataset.

- **What happened?**

The following exception was raised

```text
While parsing a protocol message, the input ended unexpectedly in the middle of a field.  This could mean either that the input has been truncated or that an embedded message misreported its own length.
```

- **What did you expect?**

The model to train.

#### Source Code

```csharp
public static IEnumerable<ImageInput> LoadImagesFromDirectory(string folder, bool useFolderNameasLabel = true)
{
    var files = Directory.GetFiles(folder, "*",
        searchOption: SearchOption.AllDirectories);

    foreach (var file in files)
    {
        if ((Path.GetExtension(file) != ".jpg") && (Path.GetExtension(file) != ".png"))
            continue;

        var label = Path.GetFileName(file);
        if (useFolderNameasLabel)
            label = Directory.GetParent(file).Name;
        else
        {
            for (int index = 0; index < label.Length; index++)
            {
                if (!char.IsLetter(label[index]))
                {
                    label = label.Substring(0, index);
                    break;
                }
            }
        }

        yield return new ImageInput()
        {
            ImagePath = file,
            Label = label
        };

    }
}
```

```csharp
MLContext mlContext = new MLContext();

IEnumerable<ImageInput> train = LoadImagesFromDirectory(trainRelativePath, true).Take(10).ToArray();
IEnumerable<ImageInput> test = LoadImagesFromDirectory(testRelativePath, true).Take(10).ToArray();

IDataView trainSet = mlContext.Data.LoadFromEnumerable(train);
IDataView testSet = mlContext.Data.LoadFromEnumerable(test);

var mapLabelTransform = mlContext.Transforms.Conversion.MapValueToKey
  (outputColumnName: "LabelAsKey",
   inputColumnName: "Label",
   keyOrdinality: ValueToKeyMappingEstimator.KeyOrdinality.ByValue);

var trainingPipeline = 
    mapLabelTransform
   .Append(mlContext.Model.ImageClassification(
       "ImagePath",
       "LabelAsKey",
       arch: ImageClassificationEstimator.Architecture.ResnetV2101,
       epoch: 100,
       batchSize: 150,
       metricsCallback: (metrics) => Console.WriteLine(metrics)));

ITransformer trainedModel = trainingPipeline.Fit(trainSet);
```

#### Logs

```text
System.FormatException
  HResult=0x80131537
  Message=Tensorflow exception triggered while loading model.
  Source=Microsoft.ML.Dnn
  StackTrace:
   at Microsoft.ML.Transforms.Dnn.DnnUtils.LoadTFSessionByModelFilePath(IExceptionContext ectx, String modelFile, Boolean metaGraph)
   at Microsoft.ML.DnnCatalog.ImageClassification(ModelOperationsCatalog catalog, String featuresColumnName, String labelColumnName, String scoreColumnName, String predictedLabelColumnName, Architecture arch, Int32 epoch, Int32 batchSize, Single learningRate, ImageClassificationMetricsCallback metricsCallback, Int32 statisticFrequency, DnnFramework framework, String modelSavePath, String finalModelPrefix, IDataView validationSet, Boolean testOnTrainSet, Boolean reuseTrainSetBottleneckCachedValues, Boolean reuseValidationSetBottleneckCachedValues, String trainSetBottleneckCachedValuesFilePath, String validationSetBottleneckCachedValuesFilePath)
   at ImageClassificationAPIMLNETSample.Program.Main(String[] args) in C:\Users\luquinta.REDMOND\source\repos\ImageClassificationAPIMLNETSample\ImageClassificationAPIMLNETSample\Program.cs:line 59

Inner Exception 1:
InvalidProtocolBufferException: While parsing a protocol message, the input ended unexpectedly in the middle of a field.  This could mean either that the input has been truncated or that an embedded message misreported its own length.
```

Additional output to the console:

```text
Google.Protobuf.InvalidProtocolBufferException: While parsing a protocol message, the input ended unexpectedly in the middle of a field.  This could mean either that the input has been truncated or that an embedded message misreported its own length.
   at Google.Protobuf.CodedInputStream.RefillBuffer(Boolean mustSucceed)
   at Google.Protobuf.CodedInputStream.ReadRawBytes(Int32 size)
   at Google.Protobuf.CodedInputStream.ReadBytes()
   at Tensorflow.TensorProto.MergeFrom(CodedInputStream input)
   at Google.Protobuf.CodedInputStream.ReadMessage(IMessage builder)
   at Tensorflow.AttrValue.MergeFrom(CodedInputStream input)
   at Google.Protobuf.CodedInputStream.ReadMessage(IMessage builder)
   at Google.Protobuf.FieldCodec.<>c__DisplayClass16_0`1.<ForMessage>b__0(CodedInputStream input)
   at Google.Protobuf.Collections.MapField`2.Codec.MessageAdapter.MergeFrom(CodedInputStream input)
   at Google.Protobuf.CodedInputStream.ReadMessage(IMessage builder)
   at Google.Protobuf.Collections.MapField`2.AddEntriesFrom(CodedInputStream input, Codec codec)
   at Tensorflow.NodeDef.MergeFrom(CodedInputStream input)
   at Google.Protobuf.CodedInputStream.ReadMessage(IMessage builder)
   at Google.Protobuf.FieldCodec.<>c__DisplayClass16_0`1.<ForMessage>b__0(CodedInputStream input)
   at Google.Protobuf.Collections.RepeatedField`1.AddEntriesFrom(CodedInputStream input, FieldCodec`1 codec)
   at Tensorflow.GraphDef.MergeFrom(CodedInputStream input)
   at Google.Protobuf.CodedInputStream.ReadMessage(IMessage builder)
   at Tensorflow.MetaGraphDef.MergeFrom(CodedInputStream input)
   at Google.Protobuf.MessageExtensions.MergeFrom(IMessage message, Byte[] data)
   at Google.Protobuf.MessageParser`1.ParseFrom(Byte[] data)
   at Tensorflow.saver._import_meta_graph_with_return_elements(String meta_graph_or_file, Boolean clear_devices, String import_scope, String[] return_elements)
   at Microsoft.ML.Transforms.Dnn.DnnUtils.<>c__DisplayClass5_0.<LoadMetaGraph>b__0(Graph graph)
   at Tensorflow.Python.tf_with[TIn,TOut](TIn py, Func`2 action)
```

### PR DESCRIPTION
**Pull request by ashbhandare**
Previously, when the ImageClassification pipeline is run for the first time, the meta graph of the model (ResnetV2101 or InceptionV3) is downloaded, and in the subsequent runs, it is reused. If the run is interrupted while the download is in progress(eg.: by stopping), the protobuff is partially downloaded. This throws an error when this incomplete graph is attempted to be read in the subsequent runs.

This changes uses resource manager to download the meta file. Resource Manager downloads the file to a temporary file path(GUID) and then atomically moves the temporary file path to the actual path. Hence once a file path is created it is guaranteed to have all the contents (unless the user or someone goes and manipulates the file)

Fixes #4234

### CODE REVIEW

File: src/Microsoft.ML.Dnn/DnnUtils.cs [hunk 1]
```diff
@@ -89,6 +90,24 @@ internal static Session LoadTFSession(IExceptionContext ectx, byte[] modelBytes,
             }
             return new Session(graph);
         }
+        internal static void MaybeDownloadFile(Uri address, string fileName)
+        {
+            using (WebClient client = new WebClient())
```
Comment (eerhardt) [inline src/Microsoft.ML.Dnn/DnnUtils.cs]: See the remarks on https://docs.microsoft.com/en-us/dotnet/api/system.net.webclient

>  Important
> We don't recommend that you use the WebClient class for new development. Instead, use the System.Net.Http.HttpClient class.

File: src/Microsoft.ML.Dnn/DnnCatalog.cs [hunk 2]
```diff
@@ -169,31 +169,20 @@ public static class DnnCatalog
...
             {
-                if (options.Arch == Architecture.InceptionV3)
-                {
-                    var baseGitPath = @"https://raw.githubusercontent.com/SciSharp/TensorFlow.NET/master/graph/InceptionV3.meta";
-                    using (WebClient client = new WebClient())
-                    {
-                        client.DownloadFile(new Uri($"{baseGitPath}"), @"InceptionV3.meta");
-                    }
+                var baseGitPath = @"https://raw.githubusercontent.com/SciSharp/TensorFlow.NET/master/graph/InceptionV3.meta";
+                DnnUtils.MaybeDownloadFile(new Uri($"{baseGitPath}"), @"InceptionV3.meta");
```
Comment (codemzs) [inline src/Microsoft.ML.Dnn/DnnCatalog.cs]: >MaybeDownloadFile [](start = 25, length = 17)

DownloadIfNeeded

File: src/Microsoft.ML.Dnn/DnnUtils.cs [hunk 3]
```diff
@@ -89,6 +90,24 @@ internal static Session LoadTFSession(IExceptionContext ectx, byte[] modelBytes,
             }
             return new Session(graph);
         }
+        internal static void MaybeDownloadFile(Uri address, string fileName)
```
Comment (codemzs) [inline src/Microsoft.ML.Dnn/DnnUtils.cs]: >MaybeDownloadFile [](start = 29, length = 17)

DownloadIfNeeded

File: src/Microsoft.ML.Dnn/DnnUtils.cs [hunk 4]
```diff
@@ -89,6 +90,24 @@ internal static Session LoadTFSession(IExceptionContext ectx, byte[] modelBytes,
...
+                if (File.Exists(fileName))
+                {
+                    client.OpenRead(address);
+                    var totalSizeInBytes = Convert.ToInt64(client.ResponseHeaders["Content-Length"]);
+                    var currentSize = new FileInfo(fileName).Length;
+
+                    //If current file size is not equal to expected file size, re-download file
+                    if (currentSize != totalSizeInBytes)
+                        client.DownloadFile(new Uri($"{address}"), fileName);
+                }
```
Comment (codemzs) [inline src/Microsoft.ML.Dnn/DnnUtils.cs]: Lets write a test for this:
1) run code that downloads the meta file via running DNN training pipeline ... make sure it trains and you get expected accuracy.
2) modify the file by copying to new file such that you copy all bytes except for few so that size is different. Delete the original file and rename this file to the older file name so that next time DNN training runs the code tries to access the "incomplete" meta file and lines 97-106 get executed.
3) re-run this piece of code via running DNN training pipeline.

Comment (ashbhandare) [reply src/Microsoft.ML.Dnn/DnnUtils.cs]: I wrote a test that still will hit the line 105, but without step 1 to cut down on test time. 

---
In reply to: [333237915](https://github.com/dotnet/machinelearning/pull/4314#discussion_r333237915) [](ancestors = 333237915)

Comment (codemzs) [reply src/Microsoft.ML.Dnn/DnnUtils.cs]: Thanks, looks [great](https://codecov.io/gh/dotnet/machinelearning/pull/4314/diff#D1-102)

File: src/Microsoft.ML.Dnn/DnnCatalog.cs [hunk 5]
```diff
@@ -169,31 +169,20 @@ public static class DnnCatalog
...
+            if (options.Arch == Architecture.InceptionV3)
             {
-                if (options.Arch == Architecture.InceptionV3)
-                {
-                    var baseGitPath = @"https://raw.githubusercontent.com/SciSharp/TensorFlow.NET/master/graph/InceptionV3.meta";
-                    using (WebClient client = new WebClient())
-                    {
-                        client.DownloadFile(new Uri($"{baseGitPath}"), @"InceptionV3.meta");
-                    }
+                var baseGitPath = @"https://raw.githubusercontent.com/SciSharp/TensorFlow.NET/master/graph/InceptionV3.meta";
```
Comment (codemzs) [inline src/Microsoft.ML.Dnn/DnnCatalog.cs]: >ar baseGitPath = @"https://raw.githubusercontent.com/SciSharp/TensorFlow.NET/master/graph/InceptionV3.meta"; [](start = 17, length = 108)

It seems we don't have a test for Inception V3 and I will prefer we add one since this piece of code seems to be doing more than just downloading meta file, it is also download check point file. It should be pretty straight forward .. just copy paste,.

File: src/Microsoft.ML.Dnn/DnnUtils.cs [hunk 6]
```diff
@@ -89,6 +92,35 @@ internal static Session LoadTFSession(IExceptionContext ectx, byte[] modelBytes,
...
+                    var totalSizeInBytes = headerResponse.Content.Headers.ContentLength;
+                    var currentSize = new FileInfo(fileName).Length;
+
+                    //If current file size is not equal to expected file size, re-download file
+                    if (currentSize != totalSizeInBytes)
+                    {
+                        File.Delete(fileName);
+                        var response = client.GetAsync(address).Result;
+                        using FileStream fileStream = new FileStream(fileName, FileMode.Create, FileAccess.Write, FileShare.None);
+                            using Stream contentStream = response.Content.ReadAsStreamAsync().Result;
```
Comment (codemzs) [inline src/Microsoft.ML.Dnn/DnnUtils.cs]: >using Stream contentStream = response.Content.ReadAsStreamAsync().Result; [](start = 28, length = 73)

align this line with 110, this is the pattern we follow, this isn't some loop :)

File: src/Microsoft.ML.Dnn/DnnUtils.cs [hunk 7]
```diff
@@ -89,6 +92,35 @@ internal static Session LoadTFSession(IExceptionContext ectx, byte[] modelBytes,
...
+                                contentStream.CopyTo(fileStream);
+                    }
+                }
+                else
+                {
+                    var response = client.GetAsync(address).Result;
+                    using FileStream fileStream = new FileStream(fileName, FileMode.Create, FileAccess.Write, FileShare.None);
+                        using Stream contentStream = response.Content.ReadAsStreamAsync().Result;
```
Comment (codemzs) [inline src/Microsoft.ML.Dnn/DnnUtils.cs]: >using Stream contentStream = response.Content.ReadAsStreamAsync().Result; [](start = 24, length = 73)

align this line with 118

File: test/Microsoft.ML.Tests/ScenariosWithDirectInstantiation/TensorflowTests.cs [hunk 8]
```diff
@@ -1518,6 +1518,91 @@ public void TensorFlowImageClassificationEarlyStoppingDecreasing()
...
+            using (File.Create(@"resnet_v2_101_299.meta")) { } 
+
+            //Create pipeline and run
+            var pipeline = mlContext.Transforms.LoadImages("Image", fullImagesetFolderPath, false, "ImagePath") // false indicates we want the image as a VBuffer<byte>
+                .Append(mlContext.Model.ImageClassification(
+                    "Image", "Label",
+                    // Just by changing/selecting InceptionV3 here instead of 
+                    // ResnetV2101 you can try a different architecture/pre-trained 
+                    // model. 
+                    arch: ImageClassificationEstimator.Architecture.ResnetV2101,
```
Comment (codemzs) [inline test/Microsoft.ML.Tests/ScenariosWithDirectInstantiation/TensorflowTests.cs]: > ImageClassificationEstimator.Architecture.ResnetV2101, [](start = 25, length = 55)

please add another test for InceptionV3, also there is a way to run test twice using "theory" with different arguments. See [https://stackoverflow.com/questions/22373258/difference-between-fact-and-theory-xunit-net](https://stackoverflow.com/questions/22373258/difference-between-fact-and-theory-xunit-net)

**Comment by codemzs** [comment]
@ashbhandare Change looks great, just some minor comments and thanks for adding this tests that covers the scenario user was hitting even though it is a negative test but important one!

**Comment by ashbhandare** [comment]
This error cannot be consistently reproduced. I am working on finding the cause.

---
In reply to: [543197416](https://github.com/dotnet/machinelearning/pull/4314#issuecomment-543197416) [](ancestors = 543197416)

**Comment by codemzs** [comment]
@ashbhandare is your branch synced to master? It seems some builds are failing because it cannot find LoadImages with 4 arguments.

**Comment by codemzs** [comment]
@ashbhandare I have pushed commits into your branch to sync it with master and resolve build errors. Just giving you heads up ...

**Comment by codemzs** [comment]
@ashbhandare I realized we have a resource manager util function that is designed for this kind of scenario. It downloads the file to a temporary file path and then moves the file to actual file atomically hence when a file path is created it is guaranteed to have all the contents. I have updated the code to use that util function and also created short urls for all our meta files, this is useful incase we wanted to changed the directed URL down the road without having to change the code.

File: src/Microsoft.ML.Dnn/DnnUtils.cs [hunk 9]
```diff
@@ -266,52 +283,18 @@ private static void CreateTempDirectoryWithAcl(string folder, string identity)
...
-                {
-                    var baseGitPath = @"https://tlcresources.blob.core.windows.net/image/ResNetV250TensorFlow/resnet_v2_50_299.meta";
-                    using (WebClient client = new WebClient())
-                    {
-                        client.DownloadFile(new Uri($"{baseGitPath}"), @"resnet_v2_50_299.meta");
-                    }
-                }
-
+                DownloadIfNeeded(env, @"tfhub_modules.zip",currentDirectory,@"tfhub_modules.zip",timeout);
+                if (!Directory.Exists(@"tfhub_modules"))
```
Comment (eerhardt) [inline src/Microsoft.ML.Dnn/DnnUtils.cs]: Why download the .zip file, but then decide afterwards not to use it?

Shouldn't you check `if (!Directory.Exists(@"tfhub_modules"))` before downloaded the .zip file?

Comment (codemzs) [reply src/Microsoft.ML.Dnn/DnnUtils.cs]: yea ok 👍🏽 i’ll fix that
